src/main/java/com/github/checkstyle/regression/Main.java

```diff
@@ -34,6 +74,160 @@ private Main() {
+                .build();
+        options.addOption(stopAfterGenerateConfig);
+
+        return options;
+    }
+
+    /**
+     * Validates the parsed CLI arguments.
+     * @param args the parsed CLI arguments.
+     * @throws IllegalArgumentException the arguments are invalid
+     */
+    private static void validateArguments(Arguments args) {
+        if (!existAndIsDirectory(args.repoPath())) {
+            throw new IllegalArgumentException(
+                    "path of local git repo must exist and be a directory");
+        }
+        if (!args.stopAfterGenerateConfig() && !existAndIsDirectory(args.testerPath().get())) {
+            throw new IllegalArgumentException(
+                    "path of checkstyle tester must exist and be a directory");
+        }
+    }
+
+    /**
+     * Runs the regression tool.
+     * @param args the parsed CLI arguments.
+     * @throws Exception excute failure
+     */
+    private static void runRegression(Arguments args) throws Exception {
+        final File config = generateConfig(args.repoPath(), args.branch());
+        System.out.println("config generated at " + config.getAbsolutePath());
+        if (!args.stopAfterGenerateConfig()) {
+            final File report = ReportGenerator.generate(
+                    args.testerPath().get(), args.repoPath(), args.branch(), config);
+            System.out.println("report generated at " + report.getAbsolutePath());
+        }
+    }
+
+    /**
+     * Checks whether the file in given path exists and is a directory.
+     * @param path the path to check
+     * @return true if file exists and is a directory
+     */
+    private static boolean existAndIsDirectory(String path) {
+        final File file = new File(path);
+        return file.exists() && file.isDirectory();
+    }
+
+    /**
+     * Generates the config file.
+     * @param repoPath the path of the checkstyle repository
+     * @param branch   the name of the PR branch
+     * @return the generated config file
+     * @throws Exception generation failure
+     */
+    private static File generateConfig(String repoPath, String branch)
+            throws Exception {
+        final List<GitChange> changes = DiffParser.parse(repoPath, branch);
+        final Map<String, ModuleExtractInfo> extractInfos =
+                ExtractInfoProcessor.getModuleExtractInfos(repoPath, branch);
+        ModuleUtils.setNameToModuleExtractInfo(extractInfos);
+        final List<ModuleInfo> moduleInfos = ModuleCollector.generate(changes);
+        final DateFormat format = new SimpleDateFormat("yyyyMMddHHmmss", Locale.getDefault());
+        final String configFileName =
+                String.format("config-%s-%s.xml", branch, format.format(new Date()));
+        return ConfigGenerator.generateConfig(configFileName, moduleInfos);
+    }
+
+    /** Represents the CLI arguments. */
+    @Value.Immutable
+    /* default */ interface Arguments {
```

Comment:
Would have a PMD failure without this. `Rule:CommentDefaultAccessModifier`
https://pmd.github.io/pmd-5.4.0/pmd-java/rules/java/comments.html#CommentDefaultAccessModifier

We need `package-private` here. `private` and `protected` can't be inherited by `immutables` and `public` is not necessary. 
We have another choice that updating pmd.xml to ignore this, instead of this comment.